File: Challenges/Leetcode/74 (Medium)/Java/Solution.java
/*
  Breeland Abrams
  Leetcode 74: Search a 2D Matrix
*/

class Solution {
  public boolean binarySearch(int[] row, int target) {
    if (row.length <= 2) {
      for (int i = 0; i <= row.length - 1; i++) {
        if (row[i] == target)
          return true;
      }
      return false;
    } else if (row[(row.length - 1) / 2] > target) {
      return binarySearch(Arrays.copyOfRange(row, 0, (row.length - 1) / 2), target);
    }
    return binarySearch(Arrays.copyOfRange(row, (row.length - 1) / 2, row.length), target);
  }

  public boolean searchMatrix(int[][] matrix, int target) {
    if (matrix.length == 0)
      return false;
    else if (matrix.length == 1)
      return binarySearch(matrix[0], target);

    for (int i = 1; i <= matrix.length - 1; i++) {
      if (matrix[i][0] > target)
        return binarySearch(matrix[i - 1], target);
    }

    return binarySearch(matrix[matrix.length - 1], target);
  }
}

File: Challenges/Leetcode/74 (Medium)/Python/Solution.py
# Breeland Abrams
# Leetcode 74: Search a 2D Matrix

class Solution:   
    def binarySearch(self, row:List[int], target:int) -> bool:
        if(len(row) <= 2): return (target in row)
        elif(target < row[int((len(row)-1 )/2)]): 
            return self.binarySearch(row[0:int((len(row)-1)/2)+1], target)
        else:
            return self.binarySearch(row[int((len(row)-1)/2):], target)
        
    
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if(len(matrix) == 0): return False
        if(len(matrix) == 1): return self.binarySearch(matrix[0], target)
        
        for i in range(1, len(matrix)):
            if matrix[i][0] > target:
                return self.binarySearch(matrix[i-1], target)
            
        return self.binarySearch(matrix[len(matrix)-1], target)